### PR TITLE
Added ShoppingCart and CartItem models, dto, mappers, services and co…

### PR DESCRIPTION
**Pull request by sovisou**
…ntrollers to implement an ability to add a certain book to user's shopping cart.

### CODE REVIEW

File: src/main/java/com/onlinebookstore/controller/ShoppingCartController.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+@RestController
+@RequestMapping("/cart")
+public class ShoppingCartController {
+    private final ShoppingCartService shoppingCartService;
+
+    @GetMapping
+    @Operation(summary = "Get shopping cart by id",
+            description = "Get a shopping cart by its identifier")
+    public ShoppingCartDto getShoppingCartByUserId(Authentication authentication) {
+        Long userId = ((User) authentication.getPrincipal()).getId();
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/controller/ShoppingCartController.java]: let's create separate method to get userId from Principal

File: src/main/java/com/onlinebookstore/controller/ShoppingCartController.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    public ShoppingCartDto addItemToCart(Authentication authentication,
+                                     @RequestBody @Valid CreateCartItemRequestDto requestDto) {
+        Long userId = ((User) authentication.getPrincipal()).getId();
+        return shoppingCartService.addCartItem(userId, requestDto);
+    }
+
+    @PutMapping("items/{cartItemId}")
+    @Operation(summary = "Update category by id",
+            description = "Update category info by its identifier")
+    public CartItemDto updateCartItemsQuantity(@PathVariable Long cartItemId,
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/controller/ShoppingCartController.java]:
```suggestion
    public ShoppingCartDto updateCartItemsQuantity(@PathVariable Long cartItemId,
```

File: src/main/java/com/onlinebookstore/dto/cart/CartItemUpdateDto.java [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package com.onlinebookstore.dto.cart;
+
+import lombok.Data;
+
+@Data
+public class CartItemUpdateDto {
+    private int quantity;
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/dto/cart/CartItemUpdateDto.java:9]:
```suggestion
    @Positive
    private int quantity;
```

File: src/main/java/com/onlinebookstore/dto/cart/CreateCartItemRequestDto.java [hunk 4]
```diff
@@ -0,0 +1,13 @@
+package com.onlinebookstore.dto.cart;
+
+import jakarta.validation.constraints.Min;
+import jakarta.validation.constraints.Positive;
+import lombok.Data;
+
+@Data
+public class CreateCartItemRequestDto {
+    @Positive
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/dto/cart/CreateCartItemRequestDto.java:9]:
```suggestion
    @Positive
    @NotNull
```

File: src/main/java/com/onlinebookstore/dto/cart/CreateCartItemRequestDto.java [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+    private Long bookId;
+    @Min(1)
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/dto/cart/CreateCartItemRequestDto.java]:
```suggestion
    @Positive
```

File: src/main/java/com/onlinebookstore/dto/cart/CreateShoppingCartRequestDto.java [hunk 6]
```diff
@@ -0,0 +1,8 @@
+package com.onlinebookstore.dto.cart;
+
+import lombok.Data;
+
+@Data
+public class CreateShoppingCartRequestDto {
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/dto/cart/CreateShoppingCartRequestDto.java]: didn't find where you use this dto in Controller

File: src/main/java/com/onlinebookstore/model/CartItem.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+@Entity
+@Table(name = "items")
+@Getter
+@Setter
+public class CartItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/model/CartItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/com/onlinebookstore/model/CartItem.java [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    @JoinColumn(name = "shopping_cart_id", nullable = false)
+    private ShoppingCart shoppingCart;
+    @ManyToOne
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/model/CartItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/com/onlinebookstore/model/ShoppingCart.java [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+@SQLDelete(sql = "UPDATE carts SET is_deleted = true WHERE id=?")
+@Where(clause = "is_deleted=false")
+@Table(name = "carts")
+@Getter
+@Setter
+public class ShoppingCart {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @OneToOne
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/model/ShoppingCart.java:31]: you can also use `@MapsId`

File: src/main/java/com/onlinebookstore/model/ShoppingCart.java [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    @JoinColumn(name = "user_id", nullable = false)
+    private User user;
+    @OneToMany(mappedBy = "shoppingCart", cascade = CascadeType.ALL)
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/model/ShoppingCart.java]:
```suggestion
    @OneToMany(mappedBy = "shoppingCart", cascade = CascadeType.ALL, orphanRemoval = true)
```

File: src/main/java/com/onlinebookstore/service/impl/UserServiceImpl.java [hunk 11]
```diff
@@ -34,6 +37,10 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto)
         user.setPassword(passwordEncoder.encode(requestDto.getPassword()));
         user.setRoles(Set.of(roleRepository.findByRole(Role.RoleName.ROLE_USER)
                 .orElseThrow(() -> new EntityNotFoundException("Role USER not found"))));
-        return userMapper.toUserResponse(userRepository.save(user));
+        user = userRepository.save(user);
+        ShoppingCart shoppingCart = new ShoppingCart();
+        shoppingCart.setUser(user);
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/service/impl/UserServiceImpl.java]: replace this logic to shopping cart service

File: src/main/java/com/onlinebookstore/service/impl/UserServiceImpl.java [hunk 12]
```diff
@@ -22,6 +24,7 @@ public class UserServiceImpl implements UserService {
     private final UserMapper userMapper;
     private final PasswordEncoder passwordEncoder;
     private final RoleRepository roleRepository;
+    private final ShoppingCartRepository shoppingCartRepository;
 
     @Override
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/service/impl/UserServiceImpl.java:29]:
```suggestion
    @Override
    @Transactional
```

File: src/main/java/com/onlinebookstore/service/impl/UserServiceImpl.java [hunk 13]
```diff
@@ -34,6 +37,10 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto)
         user.setPassword(passwordEncoder.encode(requestDto.getPassword()));
         user.setRoles(Set.of(roleRepository.findByRole(Role.RoleName.ROLE_USER)
                 .orElseThrow(() -> new EntityNotFoundException("Role USER not found"))));
-        return userMapper.toUserResponse(userRepository.save(user));
+        user = userRepository.save(user);
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/service/impl/UserServiceImpl.java]:
```suggestion
        userRepository.save(user);
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/com/onlinebookstore/mapper/BookMapper.java [hunk 14]
```diff
@@ -23,4 +25,9 @@ public interface BookMapper {
     default void setCategoryIds(@MappingTarget BookDto bookDto, Book book) {
         bookDto.setCategories(book.getCategories());
     }
+
+    @Named("bookFromId")
+    default Book bookFromId(Long id, BookRepository bookRepository) {
+        return bookRepository.findById(id).orElse(null);
+    }
```
Comment (liliia-ponomarenko) [inline src/main/java/com/onlinebookstore/mapper/BookMapper.java]: Why do you need it? The purpose of the mapper is to convert one object to another. Don't complicate it by calling repository methods.

File: src/main/java/com/onlinebookstore/service/impl/ShoppingCartServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Can't get shopping cart for user with id: " + userId));
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Override
+    @Transactional
+    public ShoppingCartDto addCartItem(Long userId, CreateCartItemRequestDto requestDto) {
+        Long book = requestDto.getBookId();
```
Comment (liliia-ponomarenko) [inline src/main/java/com/onlinebookstore/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
        Long bookId = requestDto.getBookId();
```

File: src/main/java/com/onlinebookstore/service/impl/ShoppingCartServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+        Book bookById = bookRepository.findById(book).orElseThrow(
+                () -> new EntityNotFoundException("Can't get book by id: " + book));
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId).orElseThrow(
+                () -> new EntityNotFoundException(
+                        "Can't get shopping cart for user with id: " + userId));
+        CartItem cartItem = cartItemMapper.toModel(requestDto);
+        cartItem.setBook(bookById);
+        cartItem.setShoppingCart(shoppingCart);
+        shoppingCart.getCartItems().add(cartItem);
+        return getShoppingCartByUserId(userId);
```
Comment (liliia-ponomarenko) [inline src/main/java/com/onlinebookstore/service/impl/ShoppingCartServiceImpl.java]: the save method call will be more readable

File: src/main/resources/db/changelog/changes/09-create-shopping-cart-table.yaml [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+                  constraints:
+                    nullable: false
+        - addForeignKeyConstraint:
+            baseTableName: carts
+            baseColumnNames: user_id
+            referencedTableName: users
+            referencedColumnNames: id
+            constraintName: fk_carts_user
+
+
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/09-create-shopping-cart-table.yaml]: leave only one empty line at the end of the file

File: src/main/java/com/onlinebookstore/service/impl/ShoppingCartServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+                .orElseThrow(() -> new EntityNotFoundException(
+                "Can't find cart with id: " + cartItemId));
+        cartItemMapper.updateItemFromDto(updateDto, cartItem);
+        cartItemRepository.save(cartItem);
+        ShoppingCart shoppingCart = cartItem.getShoppingCart();
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Override
+    public void removeCartItem(Long cartItemId) {
```
Comment (liliia-ponomarenko) [inline src/main/java/com/onlinebookstore/service/impl/ShoppingCartServiceImpl.java:85]: Should you check if such a CartItem exists before removing it?

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s fix a few minor comments ;)

File: src/main/java/com/onlinebookstore/service/impl/ShoppingCartServiceImpl.java [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId).orElseThrow(
+                () -> new EntityNotFoundException(
+                        "Can't get shopping cart for user with id: " + userId));
+        CartItem cartItem = cartItemMapper.toModel(requestDto);
+        cartItem.setBook(bookById);
+        cartItem.setShoppingCart(shoppingCart);
+        cartItemRepository.save(cartItem);
+        shoppingCart.getCartItems().add(cartItem);
+        shoppingCartRepository.save(shoppingCart);
+        return getShoppingCartByUserId(userId);
```
Comment (ahoienko) [inline src/main/java/com/onlinebookstore/service/impl/ShoppingCartServiceImpl.java]: What is the reason to call getShoppingCartByUserId here? Is it possible just return shopping cart?

File: src/main/java/com/onlinebookstore/service/impl/ShoppingCartServiceImpl.java (hunk 18 above)
Comment (ahoienko) [reply src/main/java/com/onlinebookstore/service/impl/ShoppingCartServiceImpl.java:85]: Not resolved

File: src/main/java/com/onlinebookstore/service/impl/ShoppingCartServiceImpl.java [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    @Override
+    public ShoppingCartDto updateCartItem(Long cartItemId, CartItemUpdateDto updateDto) {
+        CartItem cartItem = cartItemRepository.findById(cartItemId)
+                .orElseThrow(() -> new EntityNotFoundException(
+                "Can't find cart with id: " + cartItemId));
```
Comment (ahoienko) [inline src/main/java/com/onlinebookstore/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
                "Can't find cart item with id: " + cartItemId));
```

File: src/main/java/com/onlinebookstore/controller/ShoppingCartController.java [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+    public ShoppingCartDto addItemToCart(Authentication authentication,
+                                     @RequestBody @Valid CreateCartItemRequestDto requestDto) {
+        Long userId = extractUserId(authentication);
+        return shoppingCartService.addCartItem(userId, requestDto);
+    }
+
+    @PutMapping("items/{cartItemId}")
+    @Operation(summary = "Update category by id",
+            description = "Update category info by its identifier")
+    public ShoppingCartDto updateCartItemsQuantity(@PathVariable Long cartItemId,
```
Comment (ahoienko) [inline src/main/java/com/onlinebookstore/controller/ShoppingCartController.java:51]: In current implementation any user can update cart item of any other user

File: src/main/java/com/onlinebookstore/service/impl/ShoppingCartServiceImpl.java [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+
+    @Override
+    @Transactional
+    public ShoppingCartDto addCartItem(Long userId, CreateCartItemRequestDto requestDto) {
+        Long bookId = requestDto.getBookId();
+        Book bookById = bookRepository.findById(bookId).orElseThrow(
+                () -> new EntityNotFoundException("Can't get book by id: " + bookId));
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId).orElseThrow(
+                () -> new EntityNotFoundException(
+                        "Can't get shopping cart for user with id: " + userId));
```
Comment (ahoienko) [inline src/main/java/com/onlinebookstore/service/impl/ShoppingCartServiceImpl.java:63]:
```suggestion
        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId).orElse(
                createShoppingCartForUser(...
```

**Review by ahoienko — CHANGES_REQUESTED** [review]
Nice, left some comments

File: src/main/java/com/onlinebookstore/service/impl/ShoppingCartServiceImpl.java (hunk 22 above)
Comment (sovisou) [reply src/main/java/com/onlinebookstore/service/impl/ShoppingCartServiceImpl.java:63]: I have some problems with this method when trying to change the logic. 
createShoppingCartForUser has type User and 
getShoppingCartByUserId has type Long. When trying to change the logic of createShoppingCartForUser, I need to change also logic in UserServiceImpl. I tried to implement something, but it does not look good, what is the best way to proceed in this situation?

Comment (okuzan) [reply src/main/java/com/onlinebookstore/service/impl/ShoppingCartServiceImpl.java:63]: You don't need to change much
here -
ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId).orElseGet(() -> createShoppingCartForUser(user));
before the method retrieve user by id
and change method createShoppingCartForUser(User user) to return newly created shopping cart

**Review by okuzan — CHANGES_REQUESTED** [review]

**Review by liliia-ponomarenko — APPROVED** [review]
Good job! 👍

**Pull request merged**